File: leetcode/395-longest-substring-with-at-least-k-repeating-characters.py
# https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/ (top-interview-questions)

from collections import Counter

class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        if len(s) == 0 or len(s) < k:
            return 0
        if k == 1:
            return len(s)
        
        count = Counter(s) # the frequency of each character in the string

        # set of the invalid characters ( chars with freq less then k )
        # a set of characters that for sure they can not existed in any valid substring
        invalid = set() 
        for char in count:
            if count[char] < k:
                invalid.add(char)
        
        # if there is no invalid chars (the hole freq. of chars >= k) return the hole length of s
        if len(invalid) == 0: 
            return len(s)
        # if there are no valid chars (the hole freqs of chars < k) return 0
        if len(invalid) == len(s):
            return 0
        
        result = 0
        start = 0
        for i in range(len(s)):
            if s[i] in invalid:
                # if we found a char that we know for sure that is can not be included in any valid substring
                # we need to check the substring that we reached until this point using a recursion call of the func
                result = max(result, self.longestSubstring(s[start:i], k))
                start = i + 1 # we need to update our start pointer
        
        return max(result, self.longestSubstring(s[start:], k))

File: leetcode/395-longest-substring-with-at-least-k-repeating-characters.js
// https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/ (top-interview-questions)

/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var longestSubstring = function(s, k) {
  if (s.length === 0 || s.length < k) return 0;
  if (k === 1) return s.length;

  const freq = {};
  for (const c of s) {
    if(!freq.hasOwnProperty(c)) freq[c] = 0;
    freq[c] += 1;
  }

  const invalidChar = new Set(); // set of invalid charachters (freq. under k)
  for (const c in freq) {
    if (freq[c] < k) {
      invalidChar.add(c);
    }
  }

  if (invalidChar.size === 0) return s.length; // all the string chars frequency are greater than k
  if (invalidChar.size === s.length) return 0; // all the string chars frequency are less than k

  let result = 0;
  let start = 0;

  for (let i = 0; i < s.length; i += 1) {
    if (invalidChar.has(s.charAt(i))) {
      result = Math.max(result, longestSubstring(s.slice(start, i), k));
      start = i + 1;
    }
  }

  return Math.max(result, longestSubstring(s.slice(start), k));
};
